subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/OutputFileChanges.java

```diff
@@ -16,13 +16,101 @@
 
 package org.gradle.internal.execution.history.changes;
 
-import com.google.common.collect.ImmutableSortedMap;
-import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
-import org.gradle.internal.fingerprint.FileCollectionFingerprint;
+import org.gradle.internal.snapshot.CompleteFileSystemLocationSnapshot;
+import org.gradle.internal.snapshot.FileSystemSnapshot;
+import org.gradle.internal.snapshot.MissingFileSnapshot;
+import org.gradle.internal.snapshot.RootTrackingFileSystemSnapshotHierarchyVisitor;
+import org.gradle.internal.snapshot.SnapshotUtil;
+import org.gradle.internal.snapshot.SnapshotVisitResult;
 
-public class OutputFileChanges extends AbstractFingerprintChanges {
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedMap;
 
-    public OutputFileChanges(ImmutableSortedMap<String, FileCollectionFingerprint> previous, ImmutableSortedMap<String, CurrentFileCollectionFingerprint> current) {
-        super(previous, current, "Output");
+public class OutputFileChanges implements ChangeContainer {
+
+    private static final CompareStrategy.ChangeFactory<CompleteFileSystemLocationSnapshot> SNAPSHOT_CHANGE_FACTORY = new CompareStrategy.ChangeFactory<CompleteFileSystemLocationSnapshot>() {
+        @Override
+        public Change added(String path, String propertyTitle, CompleteFileSystemLocationSnapshot current) {
+            return new DescriptiveChange("Output property '%s' file %s has been added.", propertyTitle, path);
+        }
+
+        @Override
+        public Change removed(String path, String propertyTitle, CompleteFileSystemLocationSnapshot previous) {
+            return new DescriptiveChange("Output property '%s' file %s has been removed.", propertyTitle, path);
+        }
+
+        @Override
+        public Change modified(String path, String propertyTitle, CompleteFileSystemLocationSnapshot previous, CompleteFileSystemLocationSnapshot current) {
+            return new DescriptiveChange("Output property '%s' file %s has changed.", propertyTitle, path);
+        }
+    };
+
+    private static final TrivialChangeDetector.ItemComparator<CompleteFileSystemLocationSnapshot> SNAPSHOT_COMPARATOR = new TrivialChangeDetector.ItemComparator<CompleteFileSystemLocationSnapshot>() {
+        @Override
+        public boolean hasSamePath(CompleteFileSystemLocationSnapshot previous, CompleteFileSystemLocationSnapshot current) {
+            return true;
```

Comment:
Why does `hasSamePath` return always true? Shouldn't it compare the absolute paths?

Reply:
Good catch, fixed.

Reply:
You reverted the fix in https://github.com/gradle/gradle/commit/81d27828b23fce43345130b69df5ee22f533abc6#diff-8f9e720b51ddb8dc11053a4db2b0909b100fbff29895e0f0b7b4ab9c531fe54cR52 again.

Reply:
I added the check again and I added a test for `OutputFileChanges`. I think we need more tests for `OutputFileChanges`, given that the strategy isn't tested as part of `FingerPrintCompareStrategy` any more.